delphi-frontend/src/main/java/au/com/integradev/delphi/reporting/QuickFixUtils.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Sonar Delphi Plugin
+ * Copyright (C) 2023 Integrated Application Development
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package au.com.integradev.delphi.reporting;
+
+import au.com.integradev.delphi.antlr.DelphiFileStream;
+import au.com.integradev.delphi.antlr.ast.token.DelphiTokenImpl;
+import au.com.integradev.delphi.file.DelphiFile;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import org.antlr.runtime.CommonToken;
+import org.sonar.plugins.communitydelphi.api.ast.DelphiNode;
+import org.sonar.plugins.communitydelphi.api.token.DelphiToken;
+
+public class QuickFixUtils {
+  private QuickFixUtils() {
+    // Utility class
+  }
+
+  private static CommonToken getAntlrToken(DelphiToken token) {
+    return (CommonToken) ((DelphiTokenImpl) token).getAntlrToken();
+  }
+
+  public static String getNodeText(DelphiNode node) {
```

Comment:
I'm not sure what you mean by a "fundamentally stringly weakness." Quick fixes **are** replacing a text span in a file with another string. There's no getting around that by clever API design.

The "weakness", if there is one, is that the nodes in our AST offer no way to get their text representation - `getImage` is a sanitized version that is not appropriate for exact replacements.

Do you have any ideas for an alternative that would not require this?

Reply:
We also don't *have* to have this util - it will just simplify a lot of complex quick fix cases like the ones it's used for in this PR. It should be possible to get around it in most cases.

Reply:
For the use-case of replacing a text range with a particular known string, the current API is fine. From what I've seen, `sonar-java` does a **lot** of these types of trivial quick-fixes with known replacement strings, and the shape of their API seems to reflects that.

But having to explicitly know the string complicates a lot of use-cases that *should* be trivial and in turn encourages us to make our abstractions leakier. We're dropping down from "abstract syntax tree with some offset/range metadata" to "I guess it's fine to go and read the source file from application code sometimes".
So yeah, I'd say it's a weakness.

Many use-cases are simpler if we can think in terms of text ranges instead.
There should be a way to easily express "move this node/token/range to the following position" as a `DelphiTextEdit`.

Reply:
> There should be a way to easily express "move this node/token/range to the following position" as a DelphiTextEdit.

I understand that from a use case perspective there *should* be a way to do this. To do that, we need to know the exact string of something, because we can't express `move [34:12 to 34:27] to [33:2 to 33:17]` using the API that Sonar gives us. We have to express that as `replace [33:2 to [33:2] with <node text>` combined with `replace [34:12 to 34:27] with ""`.

Reply:
> This would of course still have the optimization problem that you mentioned, maybe even exacerbated if we give it first-class API support.

I'd still be nervous about performance implications of making it easy for users to trigger that file readback through a convenient API (util or not), but by abstracting the behavior in a central place we can more easily re-think/optimize the underlying implementation without being constrained by the shape and contract of some specific util function.

> I wonder if we could stop the quick fix code from running in a non-SonarLint context to mitigate the performance cost.

That's definitely an option.
We'd just need to get the `SonarProduct` into `DelphiIssueBuilderImpl` through, which probably means smuggling it through `DelphiCheckContextImpl`.

We may as well leave that alone until we do some profiling and determine that it's worth doing.

Reply:
Have a look at the new approach, see what you think.